File: net/pdfjet/Text.cs
/*
 * Text.cs
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
using System;
using System.Text;
using System.Collections.Generic;

namespace PDFjet.NET {
/// <summary>
/// Paragraphs of text lines, wrapped at a width, with an optional border.
/// Please see Example_03, Example_41 and Example_49.
/// </summary>
public class Text : IDrawable {
    private List<Paragraph> paragraphs;
    private float x1;
    private float y1;
    private float width;
    private float xText;
    private float yText;
    private float paragraphLeading = 24f;
    private bool hasBorder = false;
    private float[] borderColor;
    private float borderWidth = 0.5f;
    private String borderPattern = "[] 0";

    /// <summary>Creates a text object from the paragraphs.</summary>
    public Text(List<Paragraph> paragraphs) {
        this.paragraphs = paragraphs;
    }

    IDrawable IDrawable.SetLocation(float x, float y) {
        return SetLocation(x, y);
    }

    /// <summary>Sets the location of the top left corner of this text.</summary>
    public Text SetLocation(float x, float y) {
        this.x1 = x;
        this.y1 = y;
        return this;
    }

    /// <summary>Sets the location of the top left corner of this text.</summary>
    public Text SetLocation(double x, double y) {
        return SetLocation((float) x, (float) y);
    }

    /// <summary>Sets the width at which the lines wrap.</summary>
    public Text SetWidth(float width) {
        this.width = width;
        return this;
    }

    /// <summary>Sets the vertical distance between paragraphs.</summary>
    public Text SetParagraphLeading(float paragraphLeading) {
        this.paragraphLeading = paragraphLeading;
        return this;
    }

    /// <summary>Sets the border width.</summary>
    public Text SetBorderWidth(float borderWidth) {
        this.borderWidth = borderWidth;
        return this;
    }

    /// <summary>Sets the border color as a 0xRRGGBB value. Color.transparent removes the border.</summary>
    public Text SetBorderColor(int color) {
        if (color == Color.transparent) {
            this.borderColor = null;
            this.hasBorder = false;
            return this;
        }
        float r = ((color >> 16) & 0xff)/255f;
        float g = ((color >>  8) & 0xff)/255f;
        float b = ((color)       & 0xff)/255f;
        SetBorderColor(r, g, b);
        return this;
    }

    /// <summary>Sets the border color from red, green and blue values and draws a border around this text.</summary>
    public Text SetBorderColor(float r, float g, float b) {
        this.borderColor = new float[] {r, g, b};
        this.hasBorder = true;
        return this;
    }

    /// <summary>Sets the border color from an array of red, green and blue values and draws a border around this text.</summary>
    public Text SetBorderColor(float[] rgbColor) {
        this.borderColor = rgbColor;
        this.hasBorder = true;
        return this;
    }

    /// <summary>Sets the dash pattern of the border.</summary>
    public Text SetBorderPattern(String borderPattern) {
        this.borderPattern = borderPattern;
        return this;
    }

    /// <summary>
    /// Draws the paragraphs on the specified page. With no page nothing is drawn
    /// and the paragraphs get their coordinates.
    /// </summary>
    public float[] DrawOn(Page page) {
        TextLine firstLine = paragraphs[0].lines[0];
        this.xText = x1;
        this.yText = y1 + firstLine.font.GetAscent(firstLine.fontSize);
        foreach (Paragraph paragraph in paragraphs) {
            firstLine = paragraph.lines[0];
            paragraph.x1 = x1;
            paragraph.y1 = yText - firstLine.font.GetAscent(firstLine.fontSize);
            paragraph.xText = xText;
            paragraph.yText = yText;
            foreach (TextLine textLine in paragraph.lines) {
                float[] point = DrawTextLine(page, xText, yText, textLine);
                xText = point[0];
                yText = point[1];
                paragraph.x2 = xText;
                paragraph.y2 = yText + textLine.font.GetDescent(textLine.fontSize);
            }
            xText = x1;
            yText += paragraphLeading;
        }

        Paragraph lastParagraph = paragraphs[paragraphs.Count - 1];
        TextLine lastTextLine = lastParagraph.GetTextLines()[lastParagraph.GetTextLines().Count - 1];
        float height = ((yText - paragraphLeading) - y1) + lastTextLine.font.GetDescent(lastTextLine.fontSize);
        if (hasBorder) {
            Rect rect = new Rect(x1, y1, width, height);
            rect.SetBorderColor(this.borderColor);
            rect.SetBorderWidth(this.borderWidth);
            rect.SetBorderPattern(this.borderPattern);
            rect.DrawOn(page);
        }

        return new float[] { x1 + width, y1 + height };
    }

    // The ASCII whitespace that Java's \s matches; a no-break space does not break a line.
    private static readonly char[] whitespace = new char[] {' ', '\t', '\n', '\x0B', '\f', '\r'};

    // Draws the text line, wrapping it at the width of this text, and returns where the next text starts.
    private float[] DrawTextLine(
            Page page, float x, float y, TextLine textLine) {
        this.xText = x;
        this.yText = y;

        String[] tokens = null;
        if (StringIsCJK(textLine.text)) {
            tokens = TokenizeCJK(textLine, this.width);
        } else {
            tokens = textLine.text.Split(whitespace, StringSplitOptions.RemoveEmptyEntries);
        }

        Font font = textLine.font;
        Font fallbackFont = textLine.fallbackFont;
        float fontSize = textLine.fontSize;
        StringBuilder buf = new StringBuilder();
        foreach (String token in tokens) {
            float runLength = font.StringWidth(fallbackFont, fontSize, buf.ToString());
            float tokenWidth = font.StringWidth(fallbackFont, fontSize, token + Single.space);
            if ((runLength + tokenWidth) < (this.x1 + this.width) - this.xText) {
                buf.Append(token + Single.space);
            } else {
                new TextLine(textLine.font, buf.ToString())
                        .SetFallbackFont(textLine.GetFallbackFont())
                        .SetFontSize(textLine.GetFontSize())
                        .SetTextColor(textLine.GetTextColor())
                        .SetColorMap(textLine.GetColorMap())
                        .SetUnderline(textLine.GetUnderline())
                        .SetStrikeout(textLine.GetStrikeout())
                        .SetLanguage(textLine.GetLanguage())
                        .SetLocation(xText, yText)
                        .DrawOn(page);
                xText = x1;
                yText += textLine.GetHeight();
                buf.Length = 0;
                buf.Append(token + Single.space);
            }
        }
        new TextLine(textLine.font, buf.ToString())
                .SetFallbackFont(textLine.fallbackFont)
                .SetFontSize(textLine.GetFontSize())
                .SetTextColor(textLine.GetTextColor())
                .SetColorMap(textLine.GetColorMap())
                .SetUnderline(textLine.GetUnderline())
                .SetStrikeout(textLine.GetStrikeout())
                .SetLanguage(textLine.GetLanguage())
                .SetLocation(xText, yText)
                .DrawOn(page);

        return new float[] {xText + font.StringWidth(fallbackFont, fontSize, buf.ToString()), yText};
    }

    private bool StringIsCJK(String str) {
        // CJK Unified Ideographs Range: 4E00–9FD5
        // Hiragana Range: 3040–309F
        // Katakana Range: 30A0–30FF
        // Hangul Jamo Range: 1100–11FF
        int numOfCJK = 0;
        foreach (char ch in str) {
            if ((ch >= 0x4E00 && ch <= 0x9FD5) ||
                    (ch >= 0x3040 && ch <= 0x309F) ||
                    (ch >= 0x30A0 && ch <= 0x30FF) ||
                    (ch >= 0x1100 && ch <= 0x11FF)) {
                numOfCJK += 1;
            }
        }
        return (numOfCJK > (str.Length / 2));
    }

    private String[] TokenizeCJK(TextLine textLine, float textWidth) {
        List<String> list = new List<String>();
        StringBuilder buf = new StringBuilder();
        foreach (char ch in textLine.text) {
            if (textLine.font.StringWidth(textLine.fallbackFont, textLine.fontSize, buf.ToString() + ch) < textWidth) {
                buf.Append(ch);
            } else {
                list.Add(buf.ToString());
                buf.Length = 0;
                buf.Append(ch);
            }
        }
        if (buf.ToString().Length > 0) {
            list.Add(buf.ToString());
        }
        return list.ToArray();
    }

    /// <summary>Reads a text file and returns its paragraphs. An empty line separates the paragraphs.</summary>
    public static List<Paragraph> ParagraphsFromFile(Font f1, String filePath) {
        List<Paragraph> paragraphs = new List<Paragraph>();
        String contents = Content.OfTextFile(filePath);
        Paragraph paragraph = new Paragraph();
        TextLine textLine = new TextLine(f1);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < contents.Length; i++) {
            char ch = contents[i];
            // We need at least one character after the \n\n to begin new paragraph!
            if (i < (contents.Length - 2) &&
                    ch == '\n' && contents[i + 1] == '\n') {
                textLine.SetText(sb.ToString());
                paragraph.Add(textLine);
                paragraphs.Add(paragraph);
                paragraph = new Paragraph();
                textLine = new TextLine(f1);
                sb.Length = 0;
                i += 1;
            } else {
                sb.Append(ch);
            }
        }
        if (!sb.ToString().Equals("")) {
            textLine.SetText(sb.ToString());
            paragraph.Add(textLine);
            paragraphs.Add(paragraph);
        }
        return paragraphs;
    }

    /// <summary>Reads the lines of a UTF-8 text file, without carriage returns.</summary>
    public static List<String> ReadLines(String filePath) {
        List<String> lines = new List<String>();
        String contents = Content.OfTextFile(filePath);
        StringBuilder buffer = new StringBuilder();
        foreach (char ch in contents) {
            if (ch == '\n') {
                lines.Add(buffer.ToString());
                buffer.Length = 0;
            } else {
                buffer.Append(ch);
            }
        }
        if (buffer.Length > 0) {
            lines.Add(buffer.ToString());
        }
        return lines;
    }
}   // End of Text.cs
}   // End of namespace PDFjet.NET

File: src/text.go
// text.go
//
// Copyright (c) 2026 PDFjet Software
// Licensed under the MIT License. See LICENSE file in the project root.

package pdfjet

import (
	"strings"

	"github.com/edragoev1/pdfjet/v9/src/color"
	"github.com/edragoev1/pdfjet/v9/src/content"
	"github.com/edragoev1/pdfjet/v9/src/single"
)

// Text is paragraphs of text lines, wrapped at a width, with an optional border.
// Please see Example_03, Example_41 and Example_49.
type Text struct {
	paragraphs                  []*Paragraph
	x1, y1, xText, yText, width float32
	paragraphLeading            float32
	hasBorder                   bool
	borderColor                 [3]float32
	borderWidth                 float32
	borderPattern               string
}

// NewText creates a text object from the paragraphs.
func NewText(paragraphs []*Paragraph) *Text {
	text := new(Text)
	text.paragraphs = paragraphs
	text.paragraphLeading = 24.0
	text.borderColor = [3]float32{0.0, 0.0, 0.0}
	text.borderWidth = 0.5
	text.borderPattern = "[] 0"
	return text
}

// SetLocation sets the location of the top left corner of this text.
func (text *Text) SetLocation(x, y float32) Drawable {
	text.x1 = x
	text.y1 = y
	return text
}

// SetWidth sets the width at which the lines wrap.
func (text *Text) SetWidth(width float32) *Text {
	text.width = width
	return text
}

// SetParagraphLeading sets the vertical distance between paragraphs.
func (text *Text) SetParagraphLeading(paragraphLeading float32) *Text {
	text.paragraphLeading = paragraphLeading
	return text
}

// SetBorderWidth sets the width of the border.
func (text *Text) SetBorderWidth(borderWidth float32) *Text {
	text.borderWidth = borderWidth
	return text
}

// SetBorderPattern sets the dash pattern of the border, for example "[3] 0".
func (text *Text) SetBorderPattern(borderPattern string) *Text {
	text.borderPattern = borderPattern
	return text
}

// SetBorderColor sets the border color as a 0xRRGGBB value and draws a border
// around this text. color.Transparent removes the border.
func (text *Text) SetBorderColor(c int32) *Text {
	if c == color.Transparent {
		text.hasBorder = false
		return text
	}
	r := float32((c>>16)&0xff) / 255.0
	g := float32((c>>8)&0xff) / 255.0
	b := float32((c)&0xff) / 255.0
	text.SetBorderColorRGB([3]float32{r, g, b})
	return text
}

// SetBorderColorRGB sets the border color from red, green and blue values and draws a border around this text.
func (text *Text) SetBorderColorRGB(borderColor [3]float32) *Text {
	text.borderColor = borderColor
	text.hasBorder = true
	return text
}

// DrawOn draws the paragraphs on the specified page and returns the x and y
// coordinates of the bottom right corner of this text. With no page nothing is
// drawn and the paragraphs get their coordinates.
func (text *Text) DrawOn(page *Page) [2]float32 {
	firstLine := text.paragraphs[0].lines[0]
	text.xText = text.x1
	text.yText = text.y1 + firstLine.font.GetAscentAt(firstLine.fontSize)
	for _, paragraph := range text.paragraphs {
		firstLine = paragraph.lines[0]
		paragraph.x1 = text.x1
		paragraph.y1 = text.yText - firstLine.font.GetAscentAt(firstLine.fontSize)
		paragraph.xText = text.xText
		paragraph.yText = text.yText
		for _, textLine := range paragraph.lines {
			point := text.drawTextLine(page, text.xText, text.yText, textLine)
			text.xText = point[0]
			text.yText = point[1]
			paragraph.x2 = text.xText
			paragraph.y2 = text.yText + textLine.font.GetDescentAt(textLine.fontSize)
		}
		text.xText = text.x1
		text.yText += text.paragraphLeading
	}

	lastParagraph := text.paragraphs[len(text.paragraphs)-1]
	lastTextLine := lastParagraph.GetTextLines()[len(lastParagraph.GetTextLines())-1]
	height := ((text.yText - text.paragraphLeading) - text.y1) +
		lastTextLine.font.GetDescentAt(lastTextLine.fontSize)
	if text.hasBorder {
		rect := NewRect(text.x1, text.y1, text.width, height)
		rect.SetBorderColorRGB(text.borderColor)
		rect.SetBorderWidth(text.borderWidth)
		rect.SetBorderPattern(text.borderPattern)
		rect.DrawOn(page)
	}

	return [2]float32{text.x1 + text.width, text.y1 + height}
}

func (text *Text) drawTextLine(page *Page, x, y float32, textLine *TextLine) []float32 {
	text.xText = x
	text.yText = y

	var tokens []string
	if text.textIsCJK(textLine.text) {
		tokens = text.tokenizeCJK(textLine, text.width)
	} else {
		tokens = strings.FieldsFunc(textLine.text, isASCIIWhitespace)
	}

	font := textLine.font
	fallbackFont := textLine.fallbackFont
	fontSize := textLine.fontSize
	var buf strings.Builder
	for _, token := range tokens {
		lineWidth := font.StringWidthFB(fallbackFont, fontSize, buf.String())
		tokenWidth := font.StringWidthFB(fallbackFont, fontSize, token+single.Space)
		if (lineWidth + tokenWidth) < (text.x1+text.width)-text.xText {
			buf.WriteString(token)
			buf.WriteString(single.Space)
		} else {
			text.drawLine(page, textLine, buf.String())
			text.xText = text.x1
			text.yText += textLine.GetHeight()
			buf.Reset()
			buf.WriteString(token)
			buf.WriteString(single.Space)
		}
	}
	text.drawLine(page, textLine, buf.String())

	return []float32{text.xText + font.StringWidthFB(fallbackFont, fontSize, buf.String()), text.yText}
}

// drawLine draws one wrapped line of the text line at the current location,
// with the text line's font, colors and decorations.
func (text *Text) drawLine(page *Page, textLine *TextLine, str string) {
	line := NewTextLine(textLine.font, str)
	line.SetFallbackFont(textLine.GetFallbackFont())
	line.SetFontSize(textLine.GetFontSize())
	line.SetTextColorRGB(textLine.GetTextColor())
	line.SetColorMap(textLine.GetColorMap())
	line.SetUnderline(textLine.GetUnderline())
	line.SetStrikeout(textLine.GetStrikeout())
	line.SetLanguage(textLine.GetLanguage())
	line.SetLocation(text.xText, text.yText)
	line.DrawOn(page)
}

func (text *Text) textIsCJK(str string) bool {
	// CJK Unified Ideographs Range: 4E00–9FD5
	// Hiragana Range: 3040–309F
	// Katakana Range: 30A0–30FF
	// Hangul Jamo Range: 1100–11FF
	numOfCJK := 0
	runes := []rune(str)
	for _, ch := range runes {
		if (ch >= 0x4E00 && ch <= 0x9FD5) ||
			(ch >= 0x3040 && ch <= 0x309F) ||
			(ch >= 0x30A0 && ch <= 0x30FF) ||
			(ch >= 0x1100 && ch <= 0x11FF) {
			numOfCJK++
		}
	}
	return numOfCJK > (len(runes) / 2)
}

func (text *Text) tokenizeCJK(textLine *TextLine, textWidth float32) []string {
	tokens := make([]string, 0)
	var sb strings.Builder
	for _, ch := range textLine.text {
		if textLine.font.StringWidthFB(textLine.fallbackFont, textLine.fontSize, sb.String()+string(ch)) < textWidth {
			sb.WriteRune(ch)
		} else {
			tokens = append(tokens, sb.String())
			sb.Reset()
			sb.WriteRune(ch)
		}
	}
	if len(sb.String()) > 0 {
		tokens = append(tokens, sb.String())
	}
	return tokens
}

// ReadLines reads the lines of a UTF-8 text file, without carriage returns.
// It exits the program if the file cannot be read.
func ReadLines(filePath string) []string {
	lines := make([]string, 0)
	var buffer strings.Builder
	for _, ch := range content.OfTextFile(filePath) {
		if ch == '\n' {
			lines = append(lines, buffer.String())
			buffer.Reset()
		} else {
			buffer.WriteRune(ch)
		}
	}
	if buffer.Len() > 0 {
		lines = append(lines, buffer.String())
	}
	return lines
}

// ParagraphsFromFile reads a text file and returns its paragraphs. An empty line separates the paragraphs.
func ParagraphsFromFile(f1 *Font, filePath string) []*Paragraph {
	paragraphs := make([]*Paragraph, 0)
	paragraph := NewParagraph()
	textLine := NewEmptyTextLine(f1)
	sb := make([]rune, 0)
	runes := []rune(content.OfTextFile(filePath))
	for i := 0; i < len(runes); i++ {
		ch := runes[i]
		// We need at least one character after the \n\n to begin new paragraph!
		if i < (len(runes)-2) &&
			ch == '\n' && runes[i+1] == '\n' {
			textLine.SetText(string(sb))
			paragraph.Add(textLine)
			paragraphs = append(paragraphs, paragraph)
			paragraph = NewParagraph()
			textLine = NewEmptyTextLine(f1)
			sb = nil
			i += 1
		} else {
			sb = append(sb, ch)
		}
	}
	if len(sb) != 0 {
		textLine.SetText(string(sb))
		paragraph.Add(textLine)
		paragraphs = append(paragraphs, paragraph)
	}
	return paragraphs
}
